vault/expiration.go

```diff
@@ -114,6 +116,58 @@ func (c *Core) stopExpiration() error {
 	return nil
 }
 
+// Tidy cleans up stale leases which are associated with invalid tokens
+func (m *ExpirationManager) Tidy() error {
+	var tidyErrors *multierror.Error
+
+	tidyFunc := func(leaseID string) {
+		le, err := m.loadEntry(leaseID)
+		if err != nil {
+			tidyErrors = multierror.Append(tidyErrors, fmt.Errorf("failed to load the lease ID %q: %v", leaseID, err))
+			return
+		}
+
+		if le == nil {
+			tidyErrors = multierror.Append(tidyErrors, fmt.Errorf("nil entry for lease ID %q: %v", leaseID, err))
+			return
+		}
+
+		if le.ClientToken == "" {
+			tidyErrors = multierror.Append(tidyErrors, fmt.Errorf("missing client token for lease ID %q: %v", leaseID, err))
+			return
+		}
+
+		saltedId := m.tokenStore.SaltID(le.ClientToken)
+
+		lock := locksutil.LockForKey(m.tokenStore.tokenLocks, le.ClientToken)
+		lock.RLock()
+		defer lock.RUnlock()
```

Comment:
This needs to be given up after the lookupSalted call or if we are trying to revoke a token it'll deadlock.